pkg/prometheus/agent/operator.go

```diff
@@ -605,10 +605,35 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 		return fmt.Errorf("synchronizing web config secret failed: %w", err)
 	}
 
-	// Create governing service if it doesn't exist.
+	// If a ServiceName is specified, check that the service exists in the namespace. If it does not exist, fail
+	// the reconciliation.
+	// If the ServiceName is not specified, create a governing service if it doesn't exist.
+	// Also, ensure that the Prometheus instance is selected by the service. If not, fail the reconciliation.
+	// TODO[mviswanathsai]: is it possible to refactor this (and other) code repetitions between the Prometheus server and agent?
 	svcClient := c.kclient.CoreV1().Services(p.Namespace)
-	if err := k8sutil.CreateOrUpdateService(ctx, svcClient, makeStatefulSetService(p, c.config)); err != nil {
-		return fmt.Errorf("synchronizing governing service failed: %w", err)
+	if p.Spec.ServiceName != nil {
+		svc, err := svcClient.Get(ctx, *p.Spec.ServiceName, metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return fmt.Errorf("service %s/%s does not exist: %w", p.Namespace, *p.Spec.ServiceName, err)
+			}
+
+			return fmt.Errorf("synchronizing service failed: %w", err)
+		}
+
+		// Check that the selectors in the service actually select the Prometheus instance.
+		svcSelector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: svc.Spec.Selector})
+		if err != nil {
+			return fmt.Errorf("failed to get label selector while synchronizing service: %w", err)
+		}
+
+		if svcSelector.Matches(labels.Set(p.Labels)) {
```

Comment:
you want to check against the pod's labels, not the Prometheus object label.

Reply:
Got it.

Reply:
this is still not correct: the Kubernetes service's selector is used to select pods based on their labels. In practice you need to use `makeSelectorLabels()` (the function could use a more expressive name, e.g. `selectorForPrometheusPods`).

Reply:
Not sure to follow here, I do use `makeSelectorLabels` and then check against the label selectors present in the service.